File: aulas/pc2021/material/CustomSort.java
// Example of custom sorting of persons
// First increasingly by age and in case of tie be alphabetical order
// ----------------------------------
// Pedro Ribeiro (DCC/FCUP) - 04/10/2019
// ----------------------------------

import java.io.*;
import java.util.*;

// Classe to store a person
class Person implements Comparable<Person> {
    public int age;
    public String name;

    Person(int a, String n) {
	age = a;
	name = n;
    }

    // Define how to compare two objectos of class Person
    // This function should return:
    //  - number < 0 if the object is smaller than "p"
    //  - number > 0 if the object is larger than "p"
    //  - zero, if the object is equal "p"
    @Override
    public int compareTo(Person p) {
	if (age < p.age) return +1;
	if (age > p.age) return -1;
	return name.compareTo(p.name);
    }
}

public class CustomSort {
    public static void main(String args[]) {
	Scanner stdin = new Scanner(System.in);

	int n    = stdin.nextInt();	
	Person v[] = new Person[n]; 
	for (int i = 0; i<n; i++)
	    v[i] = new Person(stdin.nextInt(), stdin.next());
       	
	Arrays.sort(v);
	
	for (int i = 0; i<n; i++)
	    System.out.println(v[i].age + " " + v[i].name);
    }
}

File: aulas/pc2021/material/customsort.c
// Example of custom sorting of persons
// First increasingly by age and in case of tie be alphabetical order
// ----------------------------------
// Pedro Ribeiro (DCC/FCUP) - 04/10/2019
// ----------------------------------

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_STRINGS 500
#define MAX_LENGTH  30

// Type to store a person
typedef struct {
  int age;
  char name[MAX_LENGTH+1]; // (+1) Space for '\0'
} Person;

// This function compares two elements of an array (needed for qsort)
// It should return:
//  - number < 0 if "a" is smaller than "b"
//  - number > 0 if "a" is larger than "b"
//  - zero, if "a" and "b" are equal
int comparePerson(const void *a, const void *b) {
  Person *p1 = (Person *)a;
  Person *p2 = (Person *)b;

  if (p1->age < p2->age) return -1;
  if (p1->age > p2->age) return +1;
  return strcmp(p1->name, p2->name);
}

int main() {
  int i, n;
  Person v[MAX_STRINGS];
  
  scanf("%d", &n);
  for (i=0; i<n; i++)
    scanf("%d %s", &v[i].age, v[i].name);

  qsort(v, n, sizeof(Person), comparePerson);
  
  for (i=0; i<n; i++)
    printf("%d %s\n", v[i].age, v[i].name);
  
  return 0;
}
